src/generate/register.rs

```diff
@@ -223,113 +223,77 @@ pub fn fields(
-            let description = if let Some(d) = &f.description {
-                util::respace(&util::escape_brackets(d))
-            } else {
-                "".to_owned()
-            };
-
-            Ok(F {
-                _pc_w,
-                _sc,
-                description,
-                description_with_bits,
-                pc_r,
-                _pc_r,
-                pc_w,
-                bits,
-                width,
-                access: f.access,
-                evs: &f.enumerated_values,
-                sc: Ident::new(&sc, Span::call_site()),
-                mask: 1u64.wrapping_neg() >> (64 - width),
-                name: &f.name,
-                offset: u64::from(offset),
-                ty: width.to_ty()?,
-                write_constraint: f.write_constraint.as_ref(),
-            })
+    // TODO enumeratedValues
+    for f in fields.into_iter() {
+        // TODO(AJM) - do we need to do anything with this range type?
+        let BitRange { offset, width, .. } = f.bit_range;
+        let span = Span::call_site();
+        let sc = Ident::new(&f.name.to_sanitized_snake_case(), span);
+        let pc = f.name.to_sanitized_upper_case();
+        let bits = Ident::new(if width == 1 { "bit" } else { "bits" }, span);
+        let mut description_with_bits = if width == 1 {
+            format!("Bit {}", offset)
+        } else {
+            format!("Bits {}:{}", offset, offset + width - 1)
+        };
+        let description = if let Some(d) = &f.description {
+            util::respace(&util::escape_brackets(d))
+        } else {
+            "".to_owned()
+        };
+        if !description.is_empty() {
+            description_with_bits.push_str(" - ");
+            description_with_bits.push_str(&description);
         }
-    }
-
-    let fs = fields.iter().map(F::from).collect::<Result<Vec<_>>>()?;
 
-    // TODO enumeratedValues
-    for f in &fs {
         let can_read = [Access::ReadOnly, Access::ReadWriteOnce, Access::ReadWrite]
             .contains(&access)
             && (f.access != Some(Access::WriteOnly))
             && (f.access != Some(Access::WriteOnce));
         let can_write = (access != Access::ReadOnly) && (f.access != Some(Access::ReadOnly));
 
-        let bits = &f.bits;
-        let mask = &util::hex(f.mask);
-        let offset = f.offset;
-        let rv = reset_value.map(|rv| (rv >> offset) & f.mask);
-        let fty = &f.ty;
+        let mask = 1u64.wrapping_neg() >> (64 - width);
+        let hexmask = &util::hex(mask);
+        let offset = u64::from(offset);
+        let rv = reset_value.map(|rv| (rv >> offset) & mask);
+        let fty = width.to_ty()?;
+        let evs = &f.enumerated_values;
 
         let lookup_results = lookup(
-            &f.evs,
+            evs,
             fields,
             parent,
             all_registers,
             peripheral,
             all_peripherals,
         )?;
 
-        let pc_r = &f.pc_r;
-        let mut pc_w = &f.pc_r;
```

Comment:
It's important!

Reply:
It's kind of hard to see without any comments in the code. ;) I'll have a look at the build results and try to incorporate it in my tests.

Reply:
Reader and writer must use one common `Enum_A` unless field have `enumeratedValues`: one for `read-only` and other for `write-only` (`Enum_AW`).